FRA Spinner Bug

elipe17 opened the pull request:
## Summary of Changes
- Updated the `Spinner` component to use pure html/scss to render and animate the spinner
- Removed the `spinner.svg`
- Added nested exception logic for axios on the poll submission function
  - The function now continues polling even if the GET request returns a 4xx or 5xx
- Updated tests

Pull request closes #4900 

## How to Test

1. To reproduce the error you need to modify some backend code to simulate a 4xx or 5xx issue. Update the `DataFileViewSet` class with the following code snippet:
```
def retrieve(self, request, *args, **kwargs):
        return Response(status=status.HTTP_502_BAD_GATEWAY)
```
2. Make this modification on `develop` and on this branch to test things.
3. Submit a larger FRA file (see attached)
4. If on `develop` verify you do not see any spinners displaying and only a pending state. If on this branch, verify that you see the spinners and the polling completes and shows the "still processing" message.


[TEST_FRA_XLSX.xlsx](https://github.com/user-attachments/files/20039447/TEST_FRA_XLSX.xlsx)


## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [ ] Spinner still looks and feels like svg implementation
+ [ ] Spinner render is immediate
+ [ ] Regardless of backend datafiles GET endpoint availability, polling and spinners still work

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

elipe17 committed:
- Updated spinner component to be pure htmls/scss to try and avoid a request to load the svg

elipe17 committed:
- Fix display logic

elipe17 committed:
- Update spinners so that they are synchronized across rows in the submission history table or anywhere else they get used

elipe17 committed:
- remove export

elipe17 committed:
Revert "- remove export"

This reverts commit 190ee1b3300575a13e3f67a46a099ac68e708cfc.

elipe17 committed:
Revert "- Update spinners so that they are synchronized across rows in the submission history table or anywhere else they get used"

This reverts commit 95b0c70b4b7e7585130ee23ca0ca537bdb2377cc.

raftmsohani commented on tdrs-frontend/src/assets/Reports.scss:
```diff
@@ -58,4 +58,20 @@
 
```
As discussed during office hours, my suggestion is to park this PR/issue. It is not high priority and is only impacting limited number of users (as far as I understand is only one user and that is not reproducible)

elipe17 committed:
- Add extra exception block for axios exception
- Keep polling even if axios is getting 4xx or 5xx responses

elipe17 committed:
- UPdate test

elipe17 committed:
Merge branch 'develop' into 4900-fra-spinner-bug

elipe17 commented:
@raftmsohani I was able to consistently reproduce the issue locally and added a fix. I can add back the spinner SVG if that seems better, just let me know! I also updated the instructions so that you can reproduce the issue to check that the solution handles the issue gracefully.

cc. @jtimpe @mattcoleanderson

jtimpe commented on tdrs-frontend/src/actions/fraReports.js:
```diff
@@ -182,37 +182,55 @@ export const pollFraSubmissionStatus =
+          onSuccess(response?.data)
+        }
+      } catch (axiosError) {
+        // Keep polling even if we are getting 4xx or 5xx responses.
```
if we're getting consistent 403, 404, or 500 errors from the endpoint, do we want to retry? it's unlikely the request will resolve

elipe17 replied:
Update to handle specific 4xxs to short circut

elipe17 commented:
@jtimpe @raftmsohani @mattcoleanderson @run8 

What do we think about doing something like this. The polling function has onError/onSuccess callbacks that weren't implemented. I implemented the onError callback and this is what a user would see if the got a 400, 401, 403, or the polling timed out.

cc. @lhuxraft @victoriaatraft @reitermb 

<img width="1036" alt="Screenshot 2025-05-08 at 10 00 22 AM" src="https://github.com/user-attachments/assets/ac506cd4-faf3-4256-a678-6707fb895343" />

<img width="1249" alt="Screenshot 2025-05-08 at 9 48 55 AM" src="https://github.com/user-attachments/assets/157eda3c-d57a-4c79-89b0-61e55d18b4da" />

lhuxraft commented:
> @jtimpe @raftmsohani @mattcoleanderson @run8
> 
> What do we think about doing something like this. The polling function has onError/onSuccess callbacks that weren't implemented. I implemented the onError callback and this is what a user would see if the got a 400, 401, 403, or the polling timed out.
> 
> cc. @lhuxraft @victoriaatraft @reitermb
> 
> <img alt="Screenshot 2025-05-08 at 10 00 22 AM" width="1036" src="https://private-user-images.githubusercontent.com/125676261/441765270-ac506cd4-faf3-4256-a678-6707fb895343.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY3MTczNjMsIm5iZiI6MTc0NjcxNzA2MywicGF0aCI6Ii8xMjU2NzYyNjEvNDQxNzY1MjcwLWFjNTA2Y2Q0LWZhZjMtNDI1Ni1hNjc4LTY3MDdmYjg5NTM0My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwOFQxNTExMDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03ZGRkZTFlNjFkNDgxZWUwYTI3NjdiYmI0MDIwZGViNTk2YjA2MWE0ZTcwNDMyNTIzZWZmOWZlNjg3YmM5YWJlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ohKnDyl0kbXASpbdLs-VO7M-SIoTW41OCFFY47gU808"> <img alt="Screenshot 2025-05-08 at 9 48 55 AM" width="1249" src="https://private-user-images.githubusercontent.com/125676261/441765343-157eda3c-d57a-4c79-89b0-61e55d18b4da.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY3MTczNjMsIm5iZiI6MTc0NjcxNzA2MywicGF0aCI6Ii8xMjU2NzYyNjEvNDQxNzY1MzQzLTE1N2VkYTNjLWQ1N2EtNGM3OS04OWIwLTYxZTU1ZDE4YjRkYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwOFQxNTExMDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mYmZkZjM1YjcyZmM5MTNmMzg4Y2FhNjlmZjk3OTI2MDMyNTJhNzdjNjQ1NzBmYzk3ZmY3OTJlNTM3MTIwM2MwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.__EbivvW8U5orjHrxKuIe8AGa7wvtdyCk2jo83rYtfQ">

Thanks for sharing these! Could we just do the same blanket error message for both:
- The system encountered an error, please refresh the page or press Search again.

@reitermb @victoriaatraft @kristen-shaw4 Let me know if you have thoughts on this

raftmsohani reviewed (approved):
worked as intended

elipe17 committed:
Merge branch 'develop' into 4900-fra-spinner-bug

kristen-shaw4 commented:
> The system encountered an error, please refresh the page or press Search again.

This error message makes sense to me to give the user an indication of how to proceed if an error occurs.

elipe17 commented:
Here is the new message and what the submission history state would look like if it is thrown when the polling did NOT time out.

<img width="1114" alt="Screenshot 2025-05-09 at 12 18 08 PM" src="https://github.com/user-attachments/assets/7dc45ee9-8bea-41ca-8967-bdb468f71989" />

elipe17 committed:
- Added helper function to dispatch correct redux state and trigger teh onError callback when an exception happens during polling
- Added onSuccess function to indicate parsing completion

lhuxraft commented:
> Here is the new message and what the submission history state would look like if it is thrown when the polling did NOT time out.
> 
> <img alt="Screenshot 2025-05-09 at 12 18 08 PM" width="1114" src="https://private-user-images.githubusercontent.com/125676261/442242796-7dc45ee9-8bea-41ca-8967-bdb468f71989.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY4MTE4MjIsIm5iZiI6MTc0NjgxMTUyMiwicGF0aCI6Ii8xMjU2NzYyNjEvNDQyMjQyNzk2LTdkYzQ1ZWU5LThiZWEtNDFjYS04OTY3LWJkYjQ2OGY3MTk4OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwOVQxNzI1MjJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iYTBhNDBlMGQwODI0NGE4OGZjOTBjMGM3Njk2OGZjMTlmNzdmYzk4NzM4MDVjNzg1MTc4OGZjMTk2NmU1NTIzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.2xy4sNyP-sU8Am_oFf-ZhJVE_wEB8RkvMJ0qrCTRTTY">

Looks perf to me!

reitermb commented:
> > Here is the new message and what the submission history state would look like if it is thrown when the polling did NOT time out.
> > <img alt="Screenshot 2025-05-09 at 12 18 08 PM" width="1114" src="https://private-user-images.githubusercontent.com/125676261/442242796-7dc45ee9-8bea-41ca-8967-bdb468f71989.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY4MTE4MjIsIm5iZiI6MTc0NjgxMTUyMiwicGF0aCI6Ii8xMjU2NzYyNjEvNDQyMjQyNzk2LTdkYzQ1ZWU5LThiZWEtNDFjYS04OTY3LWJkYjQ2OGY3MTk4OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUwOVQxNzI1MjJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iYTBhNDBlMGQwODI0NGE4OGZjOTBjMGM3Njk2OGZjMTlmNzdmYzk4NzM4MDVjNzg1MTc4OGZjMTk2NmU1NTIzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.2xy4sNyP-sU8Am_oFf-ZhJVE_wEB8RkvMJ0qrCTRTTY">
> 
> Looks perf to me!

Seconding consolidating these cases into that message!

elipe17 committed:
- Added test for coverage on 4xx return codes

elipe17 committed:
- Add tab index on alert for accessibility

reitermb reviewed (approved):
Ran through a11y tests with Eric; one tweak made to z-index to correct one of the error banner screenreader announcements. Looking good and should be good to go!

elipe17 committed:
Merge branch 'develop' into 4900-fra-spinner-bug

elipe17 committed:
Merge branch 'develop' into 4900-fra-spinner-bug
